File: Java/DesignPatterns/src/main/java/designPatterns/creational/singleton/Singleton.java
package designPatterns.creational.singleton;

import java.time.LocalDateTime;

public class Singleton {
  private static Singleton instance;
  private LocalDateTime date;

  private Singleton() {}

  public static Singleton getInstance() {
    if (instance == null) {
      instance = new Singleton();
      instance.date = LocalDateTime.now();
    }
    return instance;
  }

  public LocalDateTime getDate() {
    return date;
  }

  @Override
  public boolean equals(Object obj) {
    if (obj instanceof Singleton) {
      if (((Singleton) obj).getDate().equals(date)) {
        return super.equals(obj);
      }
    }
    return false;
  }
}

File: Kotlin/DesignPatterns/src/main/kotlin/creational/singleton/Singleton.kt
package creational.singleton

import java.time.LocalDateTime

object Singleton {

  private var instance: Singleton? = null
  private var _date: LocalDateTime? = null

  val date: LocalDateTime?
    get() = _date

  fun getInstance(): Singleton? {
    instance =  if (instance == null) {
      _date = LocalDateTime.now()
      Singleton
    } else {
      instance
    }
    return instance
  }

  override fun equals(obj: Any?): Boolean {
    obj as Singleton
    if (_date == obj._date) {
      if (this === obj) return true
      if (javaClass == obj?.javaClass) return true
      if (instance == obj.instance) return true
    }
    return false
  }
}